src/category_theory/subobject.lean

```diff
@@ -0,0 +1,906 @@
+thin_skeleton.map₂ F
+
+@[simp]
+lemma lower_comm (F : mono_over Y ⥤ mono_over X) :
+  to_thin_skeleton _ ⋙ lower F = F ⋙ to_thin_skeleton _ :=
+rfl
+
+/-- An adjunction between `mono_over A` and `mono_over B` gives an adjunction
+between `subobject A` and `subobject B`. -/
+def lower_adjunction {A : C} {B : D}
+  {L : mono_over A ⥤ mono_over B} {R : mono_over B ⥤ mono_over A} (h : L ⊣ R) :
+  lower L ⊣ lower R :=
+thin_skeleton.lower_adjunction _ _ h
+
+/-- An equivalence between `mono_over A` and `mono_over B` gives an equivalence
+between `subobject A` and `subobject B`. -/
+@[simps]
+def lower_equivalence {A : C} {B : D} (e : mono_over A ≌ mono_over B) : subobject A ≌ subobject B :=
+{ functor := lower e.functor,
+  inverse := lower e.inverse,
+  unit_iso :=
+  begin
+    apply eq_to_iso,
+    convert thin_skeleton.map_iso_eq e.unit_iso,
+    { exact thin_skeleton.map_id_eq.symm },
+    { exact (thin_skeleton.map_comp_eq _ _).symm },
+  end,
+  counit_iso :=
+  begin
+    apply eq_to_iso,
+    convert thin_skeleton.map_iso_eq e.counit_iso,
+    { exact (thin_skeleton.map_comp_eq _ _).symm },
+    { exact thin_skeleton.map_id_eq.symm },
+  end }
+
+section pullback
+variables [has_pullbacks C]
+
+/-- When `C` has pullbacks, a morphism `f : X ⟶ Y` induces a functor `subobject Y ⥤ subobject X`,
+by pulling back a monomorphism along `f`. -/
+def pullback (f : X ⟶ Y) : subobject Y ⥤ subobject X :=
+lower (mono_over.pullback f)
+
+lemma pullback_id (x : subobject X) : (pullback (𝟙 X)).obj x = x :=
+begin
+  apply quotient.induction_on' x,
+  intro f,
+  apply quotient.sound,
+  exact ⟨mono_over.pullback_id.app f⟩,
+end
+
+lemma pullback_comp (f : X ⟶ Y) (g : Y ⟶ Z) (x : subobject Z) :
+  (pullback (f ≫ g)).obj x = (pullback f).obj ((pullback g).obj x) :=
+begin
+  apply quotient.induction_on' x,
+  intro t,
+  apply quotient.sound,
+  refine ⟨(mono_over.pullback_comp _ _).app t⟩,
+end
+
+instance (f : X ⟶ Y) : faithful (pullback f) := {}
+
+end pullback
+
+section map
+
+/--
+We can map subobjects of `X` to subobjects of `Y`
+by post-composition with a monomorphism `f : X ⟶ Y`.
+-/
+def map (f : X ⟶ Y) [mono f] : subobject X ⥤ subobject Y :=
+lower (mono_over.map f)
+
+lemma map_id (x : subobject X) : (map (𝟙 X)).obj x = x :=
+begin
+  apply quotient.induction_on' x,
+  intro f,
+  apply quotient.sound,
+  exact ⟨mono_over.map_id.app f⟩,
+end
+
+lemma map_comp (f : X ⟶ Y) (g : Y ⟶ Z) [mono f] [mono g] (x : subobject X) :
+  (map (f ≫ g)).obj x = (map g).obj ((map f).obj x) :=
+begin
+  apply quotient.induction_on' x,
+  intro t,
+  apply quotient.sound,
+  refine ⟨(mono_over.map_comp _ _).app t⟩,
+end
+
+/-- Isomorphic objects have equivalent subobject lattices. -/
+def map_iso {A B : C} (e : A ≅ B) : subobject A ≌ subobject B :=
+lower_equivalence (mono_over.map_iso e)
+
+/-- In fact, there's a type level bijection between the subobjects of isomorphic objects. -/
+@[simps]
+def map_iso_to_equiv (e : X ≅ Y) : subobject X ≃ subobject Y :=
```

Comment:
Perhaps this should be upgraded to an order isomorphism? Relatedly I wonder if it's better to state some of the functors of subobjects as bundled monotone maps instead...

Reply:
I've upgraded this to an order_iso.

(I added a `@[mono]` lemma saying functors between preorders are monotone functions.)

I think I'd prefer to punt on switching to monotone maps for now. I've added a comment in the module doc that this may be worth revisiting. Hopefully that's okay for now?

Reply:
Sounds good!